capellambse/_native.py

```diff
@@ -0,0 +1,187 @@
+# SPDX-FileCopyrightText: Copyright DB Netz AG and the capellambse contributors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Facilities for working with a native Capella instance."""
+from __future__ import annotations
+
+__all__ = ["ARGS_CMDLINE", "NativeCapella"]
+
+import contextlib
+import errno
+import logging
+import pathlib
+import shutil
+import subprocess
+import tempfile
+import typing as t
+
+import typing_extensions as te
+
+import capellambse
+
+_LOGGER = logging.getLogger(__name__)
+
+_Runner = t.Union["_SimpleRunner", "_DockerRunner"]
+_COMMON_ARGS: t.Final = (
+    "-nosplash",
+    "-consolelog",
+    "-forceimport",
+    "-input",
+    "main_model",
+)
+_ARG_WORKSPACE: t.Final = "-data"
+_ARG_PROJECT: t.Final = "-import"
+
+ARGS_CMDLINE: t.Final = (
+    "-application",
+    "org.polarsys.capella.core.commandline.core",
+)
+
+
+class NativeCapella:
+    def __init__(
+        self,
+        model: capellambse.MelodyModel,
+        *,
+        exe: str | pathlib.Path | None = None,
+        docker: str | None = None,
+    ) -> None:
+        if bool(exe) + bool(docker) != 1:
+            raise TypeError("Must specify exactly one of 'exe' or 'docker'")
+
+        self.model = model
+        self.__context: contextlib.ExitStack | None = None
+        self.workspace: pathlib.Path | None = None
+        self.project: pathlib.Path | None = None
+
+        assert model.info.capella_version
+        if exe:
+            if isinstance(exe, str):
+                exe = exe.replace("{VERSION}", model.info.capella_version)
+            self.__runner: _Runner = _SimpleRunner(exe)
+        elif docker:
+            docker = docker.replace("{VERSION}", model.info.capella_version)
+            self.__runner = _DockerRunner(docker)
+        else:
+            assert False
+
+    def __enter__(self) -> te.Self:
+        if self.__context is not None:
+            raise RuntimeError("Cannot enter context twice")
+        self.__context = ctx = contextlib.ExitStack()
+        self.__context.__enter__()
+
+        workspace = tempfile.TemporaryDirectory(prefix="workspace.")
+        ws_path = ctx.enter_context(workspace)
+        _LOGGER.debug("Creating workspace at %s", ws_path)
+        self.workspace = pathlib.Path(ws_path)
+        project = self.model._loader.write_tmp_project_dir()
+        self.project = ctx.enter_context(project)
+        _LOGGER.debug("Project files written to %s", self.project)
+        return self
+
+    def __exit__(self, *args) -> None:
+        assert self.__context is not None
+        self.__context.__exit__(*args)
+        self.__context = self.workspace = self.project = None
```

Comment:
Unless there's a strong reason not to, I suggest transforming this to a function with a `@contextlib.contextmanager` decorator. That would provide for more clearly scoped variables (== fewer checks).

Reply:
The reason here is that we `return self` out of `__enter__`, which for one acts like a callable, but also provides the public attributes `workspace` and `project` – this is needed to copy Capella's output to somewhere safe before destroying the temporary workspace. This is a bit harder to get right if we transform it to a `@contextmanager` function (also taking into account proper type hints). We'd need some class to hold this data and the `__call__` method anyway, so I thought it would make sense to keep everything together and have the class itself act as the context manager.

Reply:
In hindsight, you're right. The benefits from clear scoping outweigh the "keep everything together".